src/com/amazon/ion/impl/IonCursorBinary.java

```diff
@@ -381,13 +416,30 @@ private static IonBufferConfiguration validate(IonBufferConfiguration configurat
             limit = alreadyReadLen;
         }
         byteBuffer = ByteBuffer.wrap(buffer, 0, configuration.getInitialBufferSize());
+        isSlowMode = true;
         refillableState = new RefillableState(
             inputStream,
             configuration.getInitialBufferSize(),
             configuration.getMaximumBufferSize()
         );
     }
 
+    /*
+     * This class contains methods with the prefix 'slow'. These methods often
```

Comment:
I know you've done a lot of profiling to end up with this implementation. What are the set of trade-offs that led to defining 'slow' vs 'quick'? Is it optimizing away the need to check for completeness? Are there other ways to mitigate that?

Reply:
Yes, 'quick' (nor 'non-slow') mode is used whenever a value is either known to be fully buffered or the input is fixed. This reduces the need to check for completeness and eliminates the need to set intra-value checkpoints or go to the data source for more data.

It is also observed that even non-fixed (refillable) data sources can make use of quick mode most of the time. That's because the underlying buffer is filled with as much data as possible each time the cursor pulls from the source. So, when a container is encountered, the cursor will enter quick mode upon stepping into that container as long as at least the number of bytes declared by the container's length are present in the buffer. Since the length includes the lengths off all child values, the cursor then remains in quick mode until stepping out of the container.

Let me know if that answers your question.

Reply:
For what it's worth, the slow-vs-quick idiom is used in several methods of the existing Ion 1.0 binary reader. However, the spaces where it has enough information to use the quick path are relatively narrow.

Reply:
I've renamed all of the previously unprefixed 'quick' methods using the 'unchecked' prefix, but I've left the 'slow'-prefixed methods as-is. I think using both prefixes will give future implementers pause before choosing one or the other without consulting the commentary (e.g. "Should I really be calling a method called 'slow'?", "Is it safe for me to call a method called 'unchecked'?") See #591